enhancements/machine-api/aws-imds-v2-support.md

```diff
@@ -0,0 +1,260 @@
+This document describes and proposes several interfaces for configuring IMDS behaviour for newly created machines via Machine API and
+during the installation procedure.
+
+## Motivation
+
+By adding mechanisms for configuring IMDS behaviour we will allow our end-users to enhance the security of their clusters and simplify Amazon [recommended best practices](https://docs.aws.amazon.com/securityhub/latest/userguide/securityhub-standards-fsbp-controls.html#fsbp-ec2-8) implementation.
+At this point, IMDSv2 enablement might be done manually via AWS console, but providing such settings on Machine API is a quite loud request from our end-users.
+
+### Goals
+
+- provide mechanisms to enable/enforce IMDSv2 for newly created machines via Machine API
+- provide mechanisms to enable/enforce IMDSv2 for master machines during installation procedure for newly created clusters
+
+### Non-Goals
+
+- perform automated migration between using IMDSv1 and IMDSv2 for existing infrastructure
+
+## Proposal
+
+ - Add new options into install-config for Othe CP installation program which will be passed as metadata service settings during virtual machines creation
+ - Extend [AWSMachineProviderConfig](https://github.com/openshift/api/blob/master/machine/v1beta1/types_awsprovider.go#L12) by adding metadata service parameters
+ - Pass metadata service parameters to AWS during new machines creation
+
+### User Stories
+
+- As an OpenShift cluster administrator, I would like to enforce IMDSv2 usage during an OCP cluster deployment procedure.
+- As an OpenShift cluster administrator, I would like to enforce IMDSv2 usage for newly created machines in my cluster via MachineSets configuration.
+- As an OpenShift cluster administrator, I would like to change IMDS interaction settings for newly created machines in my cluster.
+
+### API Extensions
+
+#### AWSMachineProviderConfig changes
+
+MachineAPI design mostly replicates parameters which [AWS API](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_InstanceMetadataOptionsRequest.html) consumes.
+
+```go
+type AWSMachineProviderConfig struct {
+    // Existing fields will not be modified
+    ...
+	MetadataServiceOptions *MetadataServiceOptions `json:"metadataServiceOptions,omitempty"`
+}
+```
+
+```go
+
+type InstanceMetadataHttpTokens string
+
+const (
+    // InstanceMetadataHttpTokensRequired enforces sending of a signed token header with any instance metadata retrieval requests
+    InstanceMetadataHttpTokensRequired = "required"
+    // InstanceMetadataHttpTokensOptional allows IMDSv1 usage along with IMDSv2
+    InstanceMetadataHttpTokensOptional = "optional"
+)
+
+// MetadataServiceOptions defines the options available to a user when configuring
+// Instance Metadata Service (IMDS) Options.
+type MetadataServiceOptions struct {
+    // HttpTokens is the state of token usage for instance metadata requests from this host.
+    // Effectively enforces IMDSv2 if set to required.
```

Comment:
What does effectively mean here? Does it, or does it not enable IMDSv2?

Reply:
IMDSv2 is always enabled but optional. `required` value there prohibits IMDSv1 usage.

Reply:
I'm worried about the naming here, it's very unclear to me what the effect here is.

Is there anything we can do to clarify this at all? 

Maybe it's just a case of explaining more detail in the comment.

```
// HTTPTokens determines whether or not the host requires the use of authentication when interacting with the 
// metadata service. When using authentication, this enforces interaction with the v2 metadata service.
```